source/mesh_deformation/fastscapecc.cc

```diff
@@ -0,0 +1,632 @@
+                    const double indx = 1+(vertex(0) - grid_extent[0].first)/dx;
+
+                    // If our x or y index isn't close to a whole number, then it's likely an artifact
+                    // from using an over-resolved quadrature rule, in that case ignore it.
+                    if (abs(indx - round(indx)) >= precision)
+                      continue;
+
+
+                    // If we're in 2D, we want to take the values and apply them to every row of X points.
+                    if (dim == 2)
+                      {
+                        for (int ys=0; ys<ny; ++ys)
+                          {
+                            // FastScape indexes from 1 to n, starting at X and Y = 0, and increases
+                            // across the X row. At the end of the row, it jumps back to X = 0
+                            // and up to the next X row in increasing Y direction. We track
+                            // this to correctly place the variables later on.
+                            // Nx*ys effectively tells us what row we are in
+                            // and then indx tells us what position in that row.
+                            const double index = round(indx)+nx*ys;
+
+                            temporary_variables[0].push_back(vertex(dim-1) - grid_extent[dim-1].second);
+                            temporary_variables[1].push_back(index-1);
+
+                            for (unsigned int i=0; i<dim; ++i)
+                              {
+                                // Always convert to m/yr for FastScape
+                                temporary_variables[i+2].push_back(vel[corner][i]*year_in_seconds);
+                              }
+                          }
+                      }
+                    // 3D case
+                    else
+                      {
+                        // Because indy only gives us the row we're in, we don't need to add 2 for the ghost node.
+                        const double indy = 1+(vertex(1) - grid_extent[1].first)/dx;
+
+                        if (abs(indy - round(indy)) >= precision)
+                          continue;
+
+                        const double index = round((indy-1))*nx+round(indx);
+
+                        temporary_variables[0].push_back(vertex(dim-1) - grid_extent[dim-1].second);   //z component
+                        temporary_variables[1].push_back(index-1);
+
+                        for (unsigned int i=0; i<dim; ++i)
+                          {
+                            temporary_variables[i+2].push_back(vel[corner][i]*year_in_seconds);
+                          }
+                      }
+                  }
+              }
+
+      // // Vector to hold the velocities that represent the change to the surface.
+      std::vector<double> V(array_size);
+
+      // Initialize the variables that will be sent to FastScape.
+      std::vector<double> h(array_size, std::numeric_limits<double>::max());
```

Comment:
Is there a reason why you set initial default values to `std::numeric_limits<double>::max()` for the elevation container?

It doesn't seem like the safest way to initialize it regarding model stability (in case any element is left to its initial value later), but I might be missing something.

Reply:
Actually, I'm not sure that you even need to create `h` and `h_old` if the reason of those two variables was to interface it with fastscapelib-fortran.

Fastscapelib-fortran has its own array of elevation values (among other internal state variables defined in a context), which you normally cannot access directly using its API (you need a copy). With Fastscapelib C++ it is different: you can define your own (xtensor) array container and pass it to the C++ API.

So maybe all you need is an `xt::xarray<double> elevation` and copy its elements to/from other `std::vector<double>` variables defined in aspect?